api/debug.go

```diff
@@ -68,3 +113,189 @@ func TraceStructLog(ctx *rpc.CallContext, db state.StateDB, header *types.Header
+	}
+	var traceConfig *TraceConfig
+	if config != nil {
+		traceConfig = &config.TraceConfig
+	}
+	return s.traceTx(ctx, tx, msg, new(TracerContext), header, statedb, evmFn, traceConfig)
+}
+
+// traceTx configures a new tracer according to the provided configuration, and
+// executes the given message in the provided environment. The return value will
+// be tracer dependent.
+func (s *PrivateDebugAPI) traceTx(ctx *rpc.CallContext, tx *types.Transaction, message Msg, txctx *TracerContext, header *types.Header, statedb state.StateDB, getEVM func(state.StateDB, *vm.Config, common.Address, *big.Int) *vm.EVM, config *TraceConfig) (interface{}, error){
+	var ( 
+		timeout = defaultTraceTimeout
+		err error
+	)
+
+	if config == nil {
+		config = &TraceConfig{}
+	}
+	if config.Timeout != nil {
+		if timeout, err = time.ParseDuration(*config.Timeout); err != nil {
+			return nil, err
+		}
+	}
+	tracer := vm.NewStructLogger(&vm.LogConfig{})
+	deadlineCtx, cancel := context.WithTimeout(ctx.Context(), timeout)
+	defer cancel()
+	evm := getEVM(statedb, &vm.Config{Tracer: tracer, Debug: true, NoBaseFee: true}, message.From(), message.GasPrice())
+
+	go func() {
+		<-deadlineCtx.Done()
+		if deadlineCtx.Err() == context.DeadlineExceeded {
+			evm.Cancel()
+		}
+	}()
+
+	statedb.SetTxContext(txctx.TxHash, txctx.TxIndex)
+	result, err := ApplyMessage(evm, message, new(GasPool).AddGas(header.GasLimit))
+	if err != nil {
+		return nil, fmt.Errorf("tracing failed: %w", err)
+	}
+
+	if evm.Cancelled() {
+		return nil, fmt.Errorf("execution timeout")
+	}
+	failed := result.Failed()
+	returnData := result.ReturnData
+	if failed && !errors.Is(result.Err, vm.ErrExecutionReverted) {
+		returnData = []byte{}
+	}
+	
+	return struct {
+		Gas         uint64           `json:"gas"`
+		Failed      bool            `json:"failed"`
+		ReturnValue hexutil.Bytes   `json:"returnValue"`
+		StructLogs  []vm.StructLog  `json:"structLogs"`
+	}{
+		Gas:         result.UsedGas, 
+		Failed:      failed,
+		ReturnValue: returnData,
+		StructLogs:  tracer.StructLogs(),
+	}, nil
+
+}
+
+func (s *PrivateDebugAPI) stateAtTransaction(ctx *rpc.CallContext, blockNrOrHash vm.BlockNumberOrHash, txIndex int) (state.StateDB, *types.Header, func(state.StateDB, *vm.Config, common.Address, *big.Int) *vm.EVM, *params.ChainConfig, error){
+	var resultState state.StateDB
+	var resultHeader *types.Header
+	var resultChainCfg *params.ChainConfig
+	var resultEvmFn evmfunc
+	err := s.evmmgr.View(blockNrOrHash, common.Address{}, ctx, func(header *types.Header, statedb state.StateDB, getEVM func(state.StateDB, *vm.Config, common.Address, *big.Int) *vm.EVM, chainCfg *params.ChainConfig) error {
+		blockHash := header.Hash()
+		if hash, ok := blockNrOrHash.Hash(); ok {
+			blockHash = hash
+		}
+		var transactions []*types.Transaction
+		err := s.storage.View(blockHash, func(tx storage.Transaction) error {
+			for i := 0; ; i++ {
+				txkey := schema.Transaction(s.chainid, blockHash.Bytes(), int64(i))
```

Comment:
Cardinal nodes don't typically store transactions. It's in the Cardinal EVM schema because that's used by the producer to know where to produced transactions to, but it's consumed by Flume, not by Cardinal.

For `debug_traceCall` I'm okay with the idea that it can only run at index 0 for a block and returns errors if other offsets are specified. If we want to support other methods like `debug_traceTransaction` or `debug_traceBock`, we'll need to be able to configure another system we can call out to (like Flume or a master) to get the transaction data.